### PR TITLE
(PUP-3968) Cache the `zypper list-updates` output

### PR DESCRIPTION
**Pull request by tampakrap**
The zypper provider calls `zypper list-updates` in order to get a list of the
packages that have newer versions. Then it checks if the resource name matches
any of the items in the list, and then applies the newer version. The problem
with the above though is that it runs list-updates for every single package
instance, while it should be run only once, as it provides all the packages
that have newer versions available.

This commit caches the list-updates output and puts its contents in a hash, so
that each package resource will query this instead of zypper directly

### CODE REVIEW

**Comment by kylog** [comment]
Thanks for the contribution! Some misc comments:
1) can this have matching spec tests?
2) would it make sense to do this in a `self.instances` method? or are there reasons not to?
3) when is the cache invalidated?

**Comment by tampakrap** [comment]
Thanks for the review.
1) Current tests ensure that when there is a newer version available for the package, it will be returned properly ('describe "when getting latest version" do' in zypper_spec.rb). In order to test this specific funcionality, we need to have two package resources and make sure that puppet will run `zypper list-updates` only once. I have no idea how to do that, hints are welcome.

Additionally, to be more precise, we could have additional tests when there are no updates available (`zypper list-updates` is empty) and when the package has no newer version available (the package is not listed under `zypper list-updates`). This is not relevant to my patch though. I tried to write them, see below:

```patch
diff --git a/spec/unit/provider/package/zypper_spec.rb b/spec/unit/provider/package/zypper_spec.rb
index 3265858..c5bef70 100755
--- a/spec/unit/provider/package/zypper_spec.rb
+++ b/spec/unit/provider/package/zypper_spec.rb
@@ -115,12 +115,31 @@ describe provider_class do
   end
 
   describe "when getting latest version" do
-    it "should return a version string with valid list-updates data from SLES11sp1" do
-      fake_data = File.read(my_fixture('zypper-list-updates-SLES11sp1.out'))
+    context "when the package has available update" do
+      it "should return a version string with valid list-updates data from SLES11sp1" do
+        fake_data = File.read(my_fixture('zypper-list-updates-SLES11sp1.out'))
+        @resource.stubs(:[]).with(:name).returns "at"
+        @provider.expects(:zypper).with("list-updates").returns fake_data
+        expect(@provider.latest).to eq("3.1.8-1069.18.2")
+      end
+    end
+
+    context "when the package is in the latest version" do
+      it "should return nil with valid list-updates data from SLES11sp1" do
+        fake_data = File.read(my_fixture('zypper-list-updates-SLES11sp1.out'))
+        @resource.stubs(:[]).with(:name).returns "zypper-log"
+        @provider.expects(:zypper).with("list-updates").returns fake_data
+        expect(@provider.latest).to eq(nil)
+      end
+    end
 
-      @resource.stubs(:[]).with(:name).returns "at"
-      @provider.expects(:zypper).with("list-updates").returns fake_data
-      expect(@provider.latest).to eq("3.1.8-1069.18.2")
+    context "when there are no updates available" do
+      it "should return nil" do
+        fake_data_empty = File.read(my_fixture('zypper-list-updates-empty.out'))
+        @resource.stubs(:[]).with(:name).returns "at"
+        @provider.expects(:zypper).with("list-updates").returns fake_data_empty
+        expect(@provider.latest).to eq(nil)
+      end
     end
   end
```
and the `zypper-list-updates-empty.out` fixture:
```
Loading repository data...
Reading installed packages...
No updates found.
```

... but I get the following output:
```
Failures:

  1) Puppet::Type::Package::ProviderZypper when getting latest version when the package is in the latest version should return nil with valid list-updates data from SLES11sp1
     Failure/Error: @provider.expects(:zypper).with("list-updates").returns fake_data
     Mocha::ExpectationError:
       not all expectations were satisfied
       unsatisfied expectations:
       - expected exactly once, not yet invoked: #<Puppet::Type::Package::ProviderZypper:0xde51110>.zypper('list-updates')
       satisfied expectations:
       - allowed any number of times, not yet invoked: Signal.trap(any_parameters)
       - allowed any number of times, invoked once: #<Mock:resource>.[](:name)
       - allowed any number of times, not yet invoked: #<Mock:resource>.command(:zypper)
       - allowed any number of times, not yet invoked: #<Mock:resource>.[](:ensure)
       - allowed any number of times, not yet invoked: #<Mock:resource>.[](:name)
       - allowed any number of times, not yet invoked: #<Mock:resource>.[](any_parameters)
     # ./spec/unit/provider/package/zypper_spec.rb:131:in `block (4 levels) in <top (required)>'

  2) Puppet::Type::Package::ProviderZypper when getting latest version when there are no updates available should return nil
     Failure/Error: expect(@provider.latest).to eq(nil)
       
       expected: nil
            got: "3.1.8-1069.18.2"
       
       (compared using ==)
     # ./spec/unit/provider/package/zypper_spec.rb:141:in `block (4 levels) in <top (required)>'
```
I am unsure though if the tests are properly written, any hints will be appreciated and I'd be happy to submit them in a different PR.

2) I believe this is not needed, as I am not touching the package resource fields in any of those methods.

3) The `zypper list-updates` output is being "cached" for each puppet agent/apply run.

**Comment by adrienthebo** [comment]
As a general FYI I've been talking with @tampakrap about this and it's loosely based on the caching done in the yum package provider.

File: lib/puppet/provider/package/zypper.rb [hunk 1]
```diff
@@ -12,6 +12,30 @@
...
+  def self.latest_package_version(zypper_instance, package)
+    if @latest_versions.nil?
+      @latest_versions = list_updates(zypper_instance)
+    end
+
+    @latest_versions[package]
+  end
+
+  def self.list_updates(zypper_instance)
+    output = zypper_instance.zypper 'list-updates'
```
Comment (adrienthebo) [inline lib/puppet/provider/package/zypper.rb]: Why is this passing `zypper_instance` and calling `zypper_instance.zypper` instead of just invoking `zypper`?

Comment (tampakrap) [inline lib/puppet/provider/package/zypper.rb]: I tried removing `zypper_instance` and one test is failing now:
```
Failures:

  1) Puppet::Type::Package::ProviderZypper when getting latest version should return a version string with valid list-updates data from SLES11sp1
     Failure/Error: expect(@provider.latest).to eq("3.1.8-1069.18.2")
       
       expected: "3.1.8-1069.18.2"
            got: nil
       
       (compared using ==)
     # ./spec/unit/provider/package/zypper_spec.rb:124:in `block (3 levels) in <top (required)>'
```

**Comment by ScottGarman** [comment]
@adrienthebo can you take a look at this?

**Comment by adrienthebo** [comment]
@tampakrap try this:

```diff
diff --git i/lib/puppet/provider/package/zypper.rb w/lib/puppet/provider/package/zypper.rb
index a7aa288..c74358c 100644
--- i/lib/puppet/provider/package/zypper.rb
+++ w/lib/puppet/provider/package/zypper.rb
@@ -12,16 +12,16 @@ Puppet::Type.type(:package).provide :zypper, :parent => :rpm do
   defaultfor :operatingsystem => [:suse, :sles, :sled, :opensuse]
   confine    :operatingsystem => [:suse, :sles, :sled, :opensuse]
 
-  def self.latest_package_version(zypper_instance, package)
+  def self.latest_package_version(package)
     if @latest_versions.nil?
-      @latest_versions = list_updates(zypper_instance)
+      @latest_versions = list_updates
     end
 
     @latest_versions[package]
   end
 
-  def self.list_updates(zypper_instance)
-    output = zypper_instance.zypper 'list-updates'
+  def self.list_updates
+    output = zypper 'list-updates'
 
     avail_updates = {}
 
@@ -96,7 +96,7 @@ Puppet::Type.type(:package).provide :zypper, :parent => :rpm do
 
   # What's the latest package version available?
   def latest
-    return self.class.latest_package_version(self, @resource[:name]) ||
+    return self.class.latest_package_version(@resource[:name]) ||
       # zypper didn't find updates, pretend the current
       # version is the latest
       @property_hash[:ensure]
diff --git i/spec/unit/provider/package/zypper_spec.rb w/spec/unit/provider/package/zypper_spec.rb
index 3265858..73389fe 100755
--- i/spec/unit/provider/package/zypper_spec.rb
+++ w/spec/unit/provider/package/zypper_spec.rb
@@ -119,7 +119,7 @@ describe provider_class do
       fake_data = File.read(my_fixture('zypper-list-updates-SLES11sp1.out'))
 
       @resource.stubs(:[]).with(:name).returns "at"
-      @provider.expects(:zypper).with("list-updates").returns fake_data
+      described_class.expects(:zypper).with("list-updates").returns fake_data
       expect(@provider.latest).to eq("3.1.8-1069.18.2")
     end
   end
```

**Comment by tampakrap** [comment]
@adrienthebo your changes in the provider work as before, but your change in the test gives me the same failures as before. I have pushed them all here https://github.com/tampakrap/puppet/tree/zypper_list_updates

**Comment by tampakrap** [comment]
Updated this PR to include @adrienthebo 's improvements

**Comment by melissa** [comment]
@tampakrap @adrienthebo and I took a look at this, and it looks like that value is getting cached when specs are run. If you clear the `@latest_versions` hash after each test runs, they will pass.

```diff
diff --git a/lib/puppet/provider/package/zypper.rb b/lib/puppet/provider/package/zypper.rb
index c74358c..5d940e5 100644
--- a/lib/puppet/provider/package/zypper.rb
+++ b/lib/puppet/provider/package/zypper.rb
@@ -20,6 +20,14 @@ Puppet::Type.type(:package).provide :zypper, :parent => :rpm do
     @latest_versions[package]
   end

+
+  # @api private
+  # Reset the latest version hash to nil
+  # needed for spec tests to clear cached value
+  def self.reset!
+    @latest_versions = nil
+  end
+
   def self.list_updates
     output = zypper 'list-updates'

diff --git a/spec/unit/provider/package/zypper_spec.rb b/spec/unit/provider/package/zypper_spec.rb
index a5276d3..5509917 100755
--- a/spec/unit/provider/package/zypper_spec.rb
+++ b/spec/unit/provider/package/zypper_spec.rb
@@ -115,6 +115,7 @@ describe provider_class do
   end

   describe "when getting latest version" do
+    after do described_class.reset! end
     context "when the package has available update" do
       it "should return a version string with valid list-updates data from SLES11sp1" do
         fake_data = File.read(my_fixture('zypper-list-updates-SLES11sp1.out'))
```

**Comment by tampakrap** [comment]
Thanks a lot @melissa your solution works fine. I have pushed the changes in this PR as well

**Comment by branan** [comment]
I seem to recall (from several years ago) that caching in providers didn't always play nice with long-running agents. I'm going to play around with caching strategies in some prototype code to make sure this will behave itself.

It's been a long time since I've done this sort of thing, so It's possible I'm chasing ghosts here, but I want to be sure.

**Comment by adrienthebo** [comment]
The Puppet agent can fork when fork is supported (https://github.com/puppetlabs/puppet/blob/master/lib/puppet/agent.rb#L43) when starting a run, and will fork when running as a long lived agent (https://github.com/puppetlabs/puppet/blob/master/lib/puppet/application/agent.rb#L414) - so there's the chance that running zypper commands on Windows might cache stale data, but in that case... whut.

That's a good point though. This provider implements a [reset!](https://github.com/puppetlabs/puppet/pull/3578/files#diff-c95f2b2e67e0ee41089707370a57dd18R18) method which can be used to clear the cached data; might be good to generalize this and invoke that method on all supporting providers after the transaction is complete. That's work for a different pull, though.

**Comment by branan** [comment]
I'm willing to say we don't support the Zypper provider on Windows.

That's good to know about the forking. The last time I did this stuff was in Puppet 2.7. Things have gotten better since then, apparently.

With @adrienthebo's information, I am :+1: on this

**Comment by branan** [comment]
This is merged in 8ff06a2033fe76507367e3f713fbc3cdd5acaf9b. I made a small amendment to add the ticket number to your second commit message.

**Comment by tampakrap** [comment]
Thanks a lot! Is it going to be included in the next 3.7.x release?

**Comment by melissa** [comment]
@tampakrap this was merged into master, so it should get into the Puppet 4.1.0 release

**Comment by tampakrap** [comment]
@melissa would it be possible to get it in 3.7.x release as well then? Should I submit it against the 3.x branch?
